src/scripts/DeployToken.ts

```diff
@@ -0,0 +1,429 @@
+function getRelativeUnixTimestamp(unixTimestamp?: number | null): bigint {
+  if (!unixTimestamp) return BigInt(0);
+
+  let vestingDuration = BigInt(0);
+  if (unixTimestamp && BigInt(unixTimestamp) > BigInt(0)) {
+    const currentTimestamp = BigInt(Math.floor(Date.now() / 1000));
+    const targetTimestamp = BigInt(unixTimestamp);
+    if (targetTimestamp > currentTimestamp) {
+      vestingDuration = targetTimestamp - currentTimestamp;
+    } else {
+      console.warn('Target timestamp is in the past, using minimum duration');
+      vestingDuration = BigInt(31 * 24 * 60 * 60); // 31 days in seconds
+    }
+  }
+  return vestingDuration;
+}
+
+async function predictToken_v3_1(
+  admin: `0x${string}`,
+  name: string,
+  symbol: string,
+  image: string,
+  metadata: string,
+  context: string,
+  supply: bigint,
+  salt: `0x${string}`,
+  initialSupplyChainId: bigint
+): Promise<`0x${string}`> {
+  const create2Salt = keccak256(
+    encodeAbiParameters(
+      [{ type: 'address' }, { type: 'bytes32' }],
+      [admin, salt]
+    )
+  );
+
+  const constructorArgs = encodeAbiParameters(
+    [
+      { type: 'string' },
+      { type: 'string' },
+      { type: 'uint256' },
+      { type: 'address' },
+      { type: 'string' },
+      { type: 'string' },
+      { type: 'string' },
+      { type: 'uint256' },
+    ],
+    [
+      name,
+      symbol,
+      supply,
+      admin,
+      image,
+      metadata,
+      context,
+      initialSupplyChainId,
+    ]
+  );
+
+  const artifact = await import('./ClankerToken.json');
+  const creationCode = artifact.bytecode.object as `0x${string}`;
+  const encodedCreationCode = encodePacked(
+    ['bytes', 'bytes'],
+    [creationCode, constructorArgs]
+  );
+  const creationCodeHash = keccak256(encodedCreationCode);
+
+  const hash = keccak256(
+    encodePacked(
+      ['uint8', 'address', 'bytes32', 'bytes32'],
+      [0xff, CLANKER_FACTORY_V3_1, create2Salt, creationCodeHash]
+    )
+  );
+
+  return `0x${hash.slice(-40)}` as `0x${string}`;
+}
+
+export async function generateSalt_v3_1(
+  admin: `0x${string}`,
+  name: string,
+  symbol: string,
+  image: string,
+  metadata: string,
+  context: string,
+  supply: bigint,
+  pairedTokenAddress: `0x${string}`,
+  initialSupplyChainId: bigint
+): Promise<{ salt: `0x${string}`; token: `0x${string}` }> {
+  const startingPoint = BigInt(
+    '0x' +
+      Array.from(crypto.getRandomValues(new Uint8Array(32)))
+        .map((b) => b.toString(16).padStart(2, '0'))
+        .join('')
+  );
+  let i = startingPoint;
+  while (true) {
+    const salt = `0x${i.toString(16).padStart(64, '0')}` as `0x${string}`;
+
+    const token = await predictToken_v3_1(
+      admin,
+      name,
+      symbol,
+      image,
+      metadata,
+      context,
+      supply,
+      salt,
+      initialSupplyChainId
+    );
+
+    const tokenNum = BigInt(token);
+    const pairedTokenNum = BigInt(pairedTokenAddress);
+
+    if (tokenNum < pairedTokenNum) {
```

Comment:
do we still need this condition? it might be nice to have comments here explaining that people can put different conditions here based on what they're interested in the address looking like